File: backend/src/pebble/services/api_keys.py
from datetime import datetime, timezone

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from pebble.models.api_key import APIKey
from pebble.utils.security import generate_api_key

VALID_SCOPES: set[str] = {
    "read:transactions",
    "read:accounts",
    "read:budgets",
    "write:budgets",
    "read:insights",
}


async def create_api_key(
    user_id: str, db: AsyncSession, *, name: str, scopes: list[str]
) -> tuple[str, APIKey]:
    unknown = set(scopes) - VALID_SCOPES
    if unknown:
        raise ValueError(f"Unknown scope(s): {sorted(unknown)}")
    raw, hashed = generate_api_key()
    key = APIKey(user_id=user_id, key_hash=hashed, name=name, scopes=list(scopes))
    db.add(key)
    await db.commit()
    await db.refresh(key)
    return raw, key


async def list_api_keys(user_id: str, db: AsyncSession) -> list[APIKey]:
    result = await db.execute(
        select(APIKey)
        .where(APIKey.user_id == user_id)
        .order_by(APIKey.created_at.desc())
    )
    return list(result.scalars().all())


async def revoke_api_key(user_id: str, key_id: str, db: AsyncSession) -> None:
    result = await db.execute(
        select(APIKey).where(APIKey.id == key_id, APIKey.user_id == user_id)
    )
    key = result.scalar_one_or_none()
    if key is None:
        from fastapi import HTTPException, status
        raise HTTPException(status.HTTP_404_NOT_FOUND, "API key not found")
    if key.revoked_at is None:
        key.revoked_at = datetime.now(timezone.utc)
        await db.commit()

File: mobile/src/api/apiKeys.ts
import { apiRequest } from "./client";

export type ApiKey = {
  id: string;
  name: string;
  scopes: string[];
  last_used_at: string | null;
  revoked_at: string | null;
  created_at: string;
};

export type CreateApiKeyResponse = {
  api_key: ApiKey;
  raw_key: string;
};

export const SCOPE_OPTIONS = [
  { id: "read:transactions", label: "Read transactions" },
  { id: "read:accounts", label: "Read account balances" },
  { id: "read:budgets", label: "Read budgets" },
  { id: "write:budgets", label: "Create / edit / delete budgets" },
  { id: "read:insights", label: "Financial health and tips" },
] as const;

export async function listApiKeys(): Promise<ApiKey[]> {
  return apiRequest<ApiKey[]>("/v1/api-keys");
}

export async function createApiKey(
  name: string,
  scopes: string[]
): Promise<CreateApiKeyResponse> {
  return apiRequest<CreateApiKeyResponse>("/v1/api-keys", {
    method: "POST",
    body: { name, scopes },
  });
}

export async function revokeApiKey(id: string): Promise<void> {
  await apiRequest<void>(`/v1/api-keys/${id}`, { method: "DELETE" });
}
